// i18next-calinga-backend/dist/index.es.js
import axios from 'axios';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

function isI18NextDefaultNamespace(optionValue) {
    return (optionValue === 'translation' ||
        (Object.prototype.toString.call(optionValue) === '[object Array]' &&
            optionValue.length === 1 &&
            optionValue[0] === 'translation'));
}
var CalingaBackend = /** @class */ (function () {
    function CalingaBackend(services, backendOptions, options) {
        this.loadPath = '{{organization}}/{{team}}/{{project}}/languages/{{language}}';
        this.localesPath = '{{organization}}/{{team}}/{{project}}/languages';
        this.init(services, backendOptions, options);
    }
    CalingaBackend.prototype.init = function (services, backendOptions, options) {
        if (!services) {
            return;
        }
        this.services = services;
        this.options = __assign(__assign({}, this.getDefaultOptions()), backendOptions);
        if (backendOptions) {
            if (isI18NextDefaultNamespace(options.ns)) {
                options.ns = backendOptions.project;
            }
            if (isI18NextDefaultNamespace(options.defaultNS)) {
                options.defaultNS = backendOptions.project;
            }
        }
        if (this.services) {
            this.loadLanguages();
        }
    };
    CalingaBackend.prototype.create = function (languages, namespace, key, fallbackValue) { };
    CalingaBackend.prototype.read = function (language, namespace, callback) {
        return __awaiter(this, void 0, void 0, function () {
            var data, etag, languageResources, cachedData, backendConnector, url, response, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        etag = '';
                        if (this.options.resources) {
                            languageResources = this.options.resources[language];
                            if (languageResources) {
                                data = languageResources[namespace];
                            }
                        }
                        if (!this.options.cache) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.options.cache.read(this.buildKey(namespace, language))];
                    case 1:
                        cachedData = _a.sent();
                        if (!cachedData) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.options.cache.read(this.buildEtagKey(namespace, language))];
                    case 2:
                        etag = _a.sent();
                        data = __assign(__assign({}, data), JSON.parse(cachedData));
                        _a.label = 3;
                    case 3:
                        callback(null, data);
                        backendConnector = this.services.backendConnector;
                        url = this.services.interpolator.interpolate(this.options.serviceBaseUrl + this.loadPath, {
                            language: language,
                            project: namespace,
                            organization: this.options.organization,
                            team: this.options.team,
                        }, language, {});
                        _a.label = 4;
                    case 4:
                        _a.trys.push([4, 10, , 11]);
                        return [4 /*yield*/, axios.get(url, {
                                validateStatus: function (status) { return status === 200 || status === 304; },
                                headers: { 'If-None-Match': etag },
                                params: { includeDrafts: this.options.includeDrafts },
                            })];
                    case 5:
                        response = _a.sent();
                        if (!(response.status === 200)) return [3 /*break*/, 9];
                        data = __assign(__assign({}, data), response.data);
                        if (!this.options.cache) return [3 /*break*/, 8];
                        return [4 /*yield*/, this.options.cache.write(this.buildEtagKey(namespace, language), response.headers['etag'])];
                    case 6:
                        _a.sent();
                        return [4 /*yield*/, this.options.cache.write(this.buildKey(namespace, language), JSON.stringify(response.data))];
                    case 7:
                        _a.sent();
                        backendConnector.loaded(language + "|" + namespace, null, data);
                        return [3 /*break*/, 9];
                    case 8:
                        backendConnector.loaded(language + "|" + namespace, null, data);
                        _a.label = 9;
                    case 9: return [3 /*break*/, 11];
                    case 10:
                        error_1 = _a.sent();
                        backendConnector.loaded(language + "|" + namespace, error_1, null);
                        this.services.logger.error('load translations failed', error_1);
                        return [3 /*break*/, 11];
                    case 11: return [2 /*return*/];
                }
            });
        });
    };
    CalingaBackend.prototype.loadLanguages = function () {
        var _this = this;
        if (this.options.devMode) {
            CalingaBackend.languages = ['cimode'];
        }
        else {
            CalingaBackend.languages = [];
        }
        var url = this.services.interpolator.interpolate(this.options.serviceBaseUrl + this.localesPath, {
            project: this.options.project,
            organization: this.options.organization,
            team: this.options.team,
        }, undefined, {});
        try {
            axios.get(url).then(function (response) {
                if (response.status === 200) {
                    var languages = response.data.map(function (l) { return l.name; });
                    if (_this.options.devMode) {
                        languages.push('cimode');
                    }
                    CalingaBackend.languages = languages;
                    if (CalingaBackend.onLanguagesChanged) {
                        CalingaBackend.onLanguagesChanged(languages);
                    }
                }
            });
        }
        catch (error) {
            this.services.logger.error('load languages failed', error);
        }
    };
    CalingaBackend.prototype.getDefaultOptions = function () {
        return {
            serviceBaseUrl: 'https://api.calinga.io/v3/',
        };
    };
    CalingaBackend.prototype.buildKey = function (namespace, language) {
        return "calinga_translations_" + namespace + "_" + language;
    };
    CalingaBackend.prototype.buildEtagKey = function (namespace, language) {
        return "calinga_etag_" + namespace + "_" + language;
    };
    CalingaBackend.type = 'backend';
    return CalingaBackend;
}());

export { CalingaBackend };
